Libraries/Opc.Ua.Security.Certificates/Extensions/X509Extensions.cs

```diff
@@ -0,0 +1,275 @@
+                var extOid = extReader.ReadObjectIdentifier();
+                bool critical = false;
+                var peekTag = extReader.PeekTag();
+                if (peekTag == boolTag)
+                {
+                    critical = extReader.ReadBoolean();
+                }
+                var data = extReader.ReadOctetString();
+                return new X509Extension(new Oid(extOid), data, critical);
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Write an extension object as ASN.1.
+        /// </summary>
+        /// <param name="writer"></param>
+        /// <param name="extension"></param>
+        public static void WriteExtension(this AsnWriter writer, X509Extension extension)
+        {
+            var etag = Asn1Tag.Sequence;
+            writer.PushSequence(etag);
+            writer.WriteObjectIdentifier(extension.Oid.Value);
+            if (extension.Critical)
+            {
+                writer.WriteBoolean(extension.Critical);
+            }
+            writer.WriteOctetString(extension.RawData);
+            writer.PopSequence(etag);
+        }
+
+        /// <summary>
+        /// Build the CRL Reason extension.
+        /// </summary>
+        public static X509Extension BuildX509CRLReason(
+            CRLReason reason
+            )
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.WriteEnumeratedValue<CRLReason>(reason);
+            return new X509Extension(Oids.CrlReasonCode, writer.Encode(), false);
+        }
+
+        /// <summary>
+        /// Build the Authority Key Identifier from an Issuer CA certificate.
+        /// </summary>
+        /// <param name="issuerCaCertificate">The issuer CA certificate</param>
+        public static X509Extension BuildAuthorityKeyIdentifier(X509Certificate2 issuerCaCertificate)
+        {
+            // force exception if SKI is not present
+            var ski = issuerCaCertificate.Extensions.OfType<X509SubjectKeyIdentifierExtension>().Single();
+            return new X509AuthorityKeyIdentifierExtension(
+                ski.SubjectKeyIdentifier.FromHexString(),
+                issuerCaCertificate.SubjectName,
+                issuerCaCertificate.GetSerialNumber());
+        }
+
+        /// <summary>
+        /// Build the CRL number.
+        /// </summary>
+        public static X509Extension BuildCRLNumber(BigInteger crlNumber)
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.WriteInteger(crlNumber);
+            return new X509Extension(Oids.CrlNumber, writer.Encode(), false);
+        }
+
+        /// <summary>
+        /// Patch serial number in a Url. byte version.
+        /// </summary>
+        public static string PatchExtensionUrl(string extensionUrl, byte[] serialNumber)
+        {
+            string serial = BitConverter.ToString(serialNumber).Replace("-", "");
```

Comment:
You wrote your own to-hex method, why not use it instead of this approach?

Reply:
done